File: app/src/main/java/com/alexkn/syntact/presentation/play/flashcard/FlashcardViewModel.java
package com.alexkn.syntact.presentation.play.flashcard;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.alexkn.syntact.domain.model.Bucket;
import com.alexkn.syntact.domain.model.cto.SolvableTranslationCto;
import com.alexkn.syntact.domain.usecase.bucket.ManageBuckets;
import com.alexkn.syntact.domain.usecase.play.ManageSolvableItems;

import java.time.Instant;
import java.util.List;

import javax.inject.Inject;

public class FlashcardViewModel extends ViewModel {

    private final ManageSolvableItems manageSolvableItems;

    private final ManageBuckets manageBuckets;

    private final Instant startTime;

    private LiveData<Bucket> bucket;

    private LiveData<List<SolvableTranslationCto>> translations;

    private Long bucketId;

    @Inject
    public FlashcardViewModel(ManageSolvableItems manageSolvableItems, ManageBuckets manageBuckets) {

        super();
        this.manageSolvableItems = manageSolvableItems;
        this.manageBuckets = manageBuckets;
        this.startTime = Instant.now();
    }

    public void init(Long bucketId) {

        this.bucketId = bucketId;
        bucket = manageBuckets.getBucket(bucketId);
        translations = manageSolvableItems.getSolvableTranslations(bucketId);
    }

    LiveData<List<SolvableTranslationCto>> getTranslations() {

        return translations;
    }

    LiveData<Bucket> getBucket() {

        return bucket;
    }
}

File: app/src/main/java/com/alexkn/syntact/presentation/flashcard/FlashcardViewModel.kt
package com.alexkn.syntact.presentation.flashcard

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.alexkn.syntact.data.model.Bucket
import com.alexkn.syntact.data.model.cto.SolvableTranslationCto
import com.alexkn.syntact.data.model.views.BucketDetail
import com.alexkn.syntact.domain.repository.BucketRepository
import com.alexkn.syntact.domain.repository.SolvableItemRepository
import com.alexkn.syntact.presentation.bucketdetails.BucketDetailsViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import java.time.Instant
import javax.inject.Inject


class FlashcardViewModel @Inject constructor(
        private val solvableItemRepository: SolvableItemRepository,
        private val bucketRepository: BucketRepository
) : ViewModel() {

    var bucket: LiveData<BucketDetail>? = null
        private set

    var translation: MutableLiveData<SolvableTranslationCto?> = MutableLiveData()
        private set

    private var bucketId: Long? = null

    fun init(bucketId: Long) {

        this.bucketId = bucketId
        bucket = bucketRepository.getBucketDetail(bucketId)
        fetchNext()
    }

    fun fetchNext() {
        translation.value = null
        viewModelScope.launch(Dispatchers.Default) {
            val nextTranslation = solvableItemRepository.findNextSolvableTranslation(bucketId!!, Instant.now())
            translation.postValue(nextTranslation)
        }
    }

    fun checkSolution(solution: String): Boolean {
        return if (translation.value?.solvableItem?.text.equals(solution, ignoreCase = true)) {
            viewModelScope.launch(Dispatchers.Default) { solvableItemRepository.solvePhrase(translation.value!!) }
            true
        } else {
            viewModelScope.launch(Dispatchers.Default) { solvableItemRepository.phraseIncorrect(translation.value!!) }
            false
        }

    }
}
